### DISCUSSION TITLE
Bengali Language Leader

### DISCUSSION THREAD

**Discussion by poltalashka** [OHF-Voice/intents#507] [Language Leaders]
About this language: 
https://en.wikipedia.org/wiki/Bengali_language

lang:bn

Got info from https://www.home-assistant.io/blog/2022/12/20/year-of-voice/

On Discord as "Polta", its new to me, and dont see a thread for "Language Leader"
I have background in IT. 

Thanks. 
Polta

**Comment by synesthesiam** [comment]
Hi @poltalashka, I've added empty Bengali (`bn`) templates to the repository. I'd be happy to have you as a language leader; please submit at least two small PRs with translations, and we can begin :slightly_smiling_face: 

By the way, I'd be interested in your feedback on some [Bengali text to speech voices I trained](https://mycroftai.github.io/mimic3-voices/#bn_multi_low).

**Reply by poltalashka** [reply]
Sure thing @synesthesiam, delighted. 
Submit 2 small paragraphs with translations: just paste it here ?

**Reply by poltalashka** [reply]
About the Bengali text to speech voices, I randomly heard 8-9 of them,  the speech it self is fine but some seems faster than normal speed, punctuation-rushed. Over all great work @synesthesiam!

Thx for the BN template.

**Reply by poltalashka** [reply]
@synesthesiam ,,, here are two sample paragraphs I translated:

Part1 (source: https://www.bbc.com/news/world-asia-41566561 : December 10th 2023): 

Myanmar Rohingya: What you need to know about the crisis

In August 2017, a deadly crackdown by Myanmar's army on Rohingya Muslims sent hundreds of thousands fleeing across the border into Bangladesh.

They risked everything to escape by sea or on foot a military offensive which the United Nations later described as a "textbook example of ethnic cleansing".


Part1 in Bangla/Bengali: 

মিয়ানমারের রোহিঙ্গা: সংকটে জ্ঞাতব্য 

২০১৭ সালের আগস্টে, রোহিঙ্গা মুসলমানদের উপর মিয়ানমারের সেনাবাহিনীর একটি মারাত্মক দমন-পীড়নের অত্যাচারের  ফলে কয়েক হাজার মানুষ সীমান্ত পেরিয়ে বাংলাদেশে পালিয়ে যায়। 

তারা প্রাণ ঝুঁকি নিয়ে সমুদ্রপথে বা পায়ে হেঁটে পালিয়ে ছিল  সামরিক আক্রমণের প্রতিকারে যা পরবর্তীতে জাতিসংঘ "জাতি-নির্মূলের পাঠ্যপুস্তকে আদর্শ " হিসাবে বর্ণনা করে।



Part2 (source: https://www.cnn.com/specials/asia/rohingya  : December 10th 2023):


The Rohingya crisis

Hundreds of thousands of Rohingya people -- a stateless Muslim minority in the country's Rakhine state -- have fled to neighboring Bangladesh following a brutal campaign of killing and arson by the Myanmar military in 2016 and 2017.

The United States government determined in 2022 the military committed genocide and crimes against humanity against the group.


Part2 in Bangla/Bengali: 

রোহিঙ্গা সংকটঃ

২০১৬ এবং ২০১৭ সালে মিয়ানমারের সেনাবাহিনীর দ্বারা হত্যা ও অগ্নিসংযোগের নৃশংস অভিযানের পর লাখ লাখ রোহিঙ্গা মানুষ -- দেশটির রাখাইন রাজ্যের একটি রাষ্ট্রবিতারিত  মুসলিম সংখ্যালঘু সম্প্রদায়  -- প্রতিবেশী বাংলাদেশে পালিয়ে গিয়েছে। 

মার্কিন যুক্তরাষ্ট্র সরকার ২০২২ সালে সিদ্ধান্ত দিয়েছিল  যে সেনাবাহিনী এই সম্প্রদায়ের  বিরুদ্ধে গণহত্যা এবং মানবতার বিরুদ্ধে অপরাধ করেছে।

**Reply by synesthesiam** [reply]
@poltalashka Sorry, I wasn't clear that PR = pull request. I would start with the [_common.yaml](https://github.com/home-assistant/intents/blob/main/sentences/bn/_common.yaml) and then proceed to [homeassistant_HassTurnOn.yaml](https://github.com/home-assistant/intents/blob/main/sentences/bn/homeassistant_HassTurnOn.yaml) and  [homeassistant_HassTurnOff.yaml](https://github.com/home-assistant/intents/blob/main/sentences/bn/homeassistant_HassTurnOff.yaml)

**Reply by poltalashka** [reply]
@synesthesiam Sure, working on it.

**Reply by poltalashka** [reply]
@synesthesiam  I am getting the following error message when I modified the "homeassistant_HassTurnOff.yaml" in stentences & tests directory.

`Running all tests for BN

Validation failed

Language: bn
 - sentences/bn/homeassistant_HassTurnOff.yaml: invalid format: Sentences should only contain words and matching syntax. They should not contain punctuation. @ data['intents']['HassTurnOff']['data'][0]['sentences'][0]. Got 'বন্ধ করুন <name>'`

**Reply by arunshekher** [reply]
I think the `SENTENCE_MATCHER` regex doesn't match all unicode words, especially the pseudo-alphabetical unicode scripts. A work around which worked for me is to make use of `expansion_rules:` in `_common.yaml`.

**Reply by synesthesiam** [reply]
We've had to [manually add punctuation](https://github.com/home-assistant/intents/blob/663bf4207b1d086fcaa2afbf20a83305a93b1cf1/script/intentfest/validate.py#L102) because the Unicode class is too broad.

Can you paste the entire sentence template here that's having problems?

**Reply by poltalashka** [reply]
Those are @synesthesiam  : 
      - sentences:
          - "বন্ধ করুন <name>"
          - "<name> বন্ধ করুন"

**Reply by poltalashka** [reply]
I see these unicode characters dont even show correctly in my last posting . Guess git container is treating `name` as variable!
They should look like these : 
![image](https://user-images.githubusercontent.com/42247336/212489472-2d586434-3df9-47fa-8e33-f23eba59a308.png)

**Reply by poltalashka** [reply]
> I think the `SENTENCE_MATCHER` regex doesn't match all unicode words, especially the pseudo-alphabetical unicode scripts. A work around which worked for me is to make use of `expansion_rules:` in `_common.yaml`.

@arunshekher Could you point me to how you implemented `expansion_rules:` in `_common.yaml`. ?

I think I found it, trying to comprehend it. 

![image](https://user-images.githubusercontent.com/42247336/212496951-c5b5efe2-0d79-4b74-98aa-a3790f644dac.png)

**Reply by arunshekher** [reply]
> We've had to [manually add punctuation](https://github.com/home-assistant/intents/blob/663bf4207b1d086fcaa2afbf20a83305a93b1cf1/script/intentfest/validate.py#L102) because the Unicode class is too broad.
> 
> Can you paste the entire sentence template here that's having problems?

@synesthesiam, I am not formally trained in this field but as far as I understand the `SENTENCE_MATCHER` regular expression is not matching ([please check](https://regex101.com/r/fWEd3M/1 )) words written in scripts originated from [south and central asian ancient scripts](https://www.unicode.org/versions/Unicode15.0.0/ch14.pdf) like _Brahmi_, _Bhaiksuki_, _Marchen_ etc collectively known as [**_Abugida_**](https://en.wikipedia.org/wiki/Abugida).

**Reply by arunshekher** [reply]
@poltalashka, I don't know if it is the expected or right way to do it but you can take a look at the language (Malayalam) I contribute to here: [main/intents/sentences/ml](https://github.com/home-assistant/intents/tree/main/sentences/ml)

sidenote: you just have to format the **markdown**  the right way to make the yaml markup show up correctly here, like so:

```yaml
      - sentences:
          - "বন্ধ করুন <name>"
          - "<name> বন্ধ করুন"
 ```

**Reply by poltalashka** [reply]
It seems in interpreting the  BN compound characters ( when two characters merges and creates a new character form, see image below) an extra step is used where certain symbols ( captures low or high emphasis) are added, these symbols are interpreted by regex as punctuation and throwing errors. `They should not contain punctuation.`

compound characters :
![image](https://user-images.githubusercontent.com/42247336/212534725-3319b1ac-dda2-4165-a8d8-5a3698f4a821.png)

**Reply by arunshekher** [reply]
@poltalashka probably yes, I think the `\w ` regex character class isn't enough to match the _**pseudo-alphabetical**_ unicode scripts.

**Reply by poltalashka** [reply]
@arunshekher, yes it seems so. I even tried using only non-compound characters but still the same issue. The `\w` simply does not work for even plan characters. Example "বনধ করউন".

**Reply by poltalashka** [reply]
Posted a PR only for common.yaml . details [https://github.com/home-assistant/intents/pull/660](url)

**Reply by synesthesiam** [reply]
OK, I can see the problem here:

```python
re.match(r"\w+", "বন্ধ")
# <re.Match object; span=(0, 2), match='বন'>
```

This seems related: https://stackoverflow.com/questions/74932163/regular-expression-cant-match-the-whole-bengali-word
I'll investigate.

**Reply by synesthesiam** [reply]
Adding `\p{M}` to the list of valid sentence characters, per the Stackoverflow suggestion. This allows the combining marks in Bengali to be matched :+1:

**Reply by arunshekher** [reply]
@synesthesiam thank you very much for working on the fix. I can confirm that it works with _Malayalam_ now, and probably works for all other _Brahmi_ derived scripts as well. Assumably this was the delay in seeing intent sentences PRs from other Indic scripts.

**Reply by synesthesiam** [reply]
You're welcome!

**Reply by poltalashka** [reply]
@synesthesiam unfortunately I see the same error still happening. 
Recreated the branch to make sure updated method is pulled, and tried both "\<name\>" & "{name}" incase that had anything to do with the error. 

ERROR message : `sentences/bn/homeassistant_HassTurnOff.yaml: invalid format: Sentences should only contain words and matching syntax. They should not contain punctuation.`

My repo branch name "mybn3main"

Thanks

**Reply by arunshekher** [reply]
@poltalashka There is changes in dependencies. Did you update the dependencies with `python3 -m pip install -r requirements.txt`

Just after the unicode regex fix I had tested `homeassistant_HassTurnOff.yaml` with one Bengali sample sentence too on a local branch and it had passed the test.

I tested again now. Take a look at this screenshot, here I'm using codespaces:
<img width="1530" alt="Screen Shot 2023-01-23 at 2 49 11 PM" src="https://user-images.githubusercontent.com/315195/214021732-33591538-2179-40f0-aaca-f46293f26558.png">

**Reply by poltalashka** [reply]
@arunshekher thanks for chiming in, I just ran the newer requirements.txt to update env libs, it updated fine (see log-details) but test failed. 

I am not sure I understand the relationship between the same yaml file sitting in "sentences" and "test" directory. I been only adding bangla words in  "sentences" dir running test from Terminal> Run Task > RunTestForCurr-File. 

![image](https://user-images.githubusercontent.com/42247336/214130202-683a13c4-c4b4-428c-a4b8-724117d8c479.png)

  


Log-Details: 
```shell
@poltalashka ➜ /workspaces/intents (mybn3main) $ cat requirements.txt 
hassil==0.2.3
PyYAML==6.0
voluptuous==0.13.1

black==22.12.0
flake8==6.0.0
isort==5.11.4
mypy==0.991
pylint==2.15.9
pytest==7.2.0
@poltalashka ➜ /workspaces/intents (mybn3main) $ ls -al requirements.txt 
-rw-rw-rw- 1 vscode root 130 Jan 16 21:44 requirements.txt
@poltalashka ➜ /workspaces/intents (mybn3main) $ python3 -m pip install -r requirements.txt
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: hassil==0.2.3 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 1)) (0.2.3)
Requirement already satisfied: PyYAML==6.0 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 2)) (6.0)
Requirement already satisfied: voluptuous==0.13.1 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 3)) (0.13.1)
Requirement already satisfied: black==22.12.0 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 5)) (22.12.0)
Requirement already satisfied: flake8==6.0.0 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 6)) (6.0.0)
Requirement already satisfied: isort==5.11.4 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 7)) (5.11.4)
Requirement already satisfied: mypy==0.991 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 8)) (0.991)
Requirement already satisfied: pylint==2.15.9 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 9)) (2.15.9)
Requirement already satisfied: pytest==7.2.0 in /home/vscode/.local/lib/python3.10/site-packages (from -r requirements.txt (line 10)) (7.2.0)
Requirement already satisfied: antlr4-python3-runtime==4.11.1 in /home/vscode/.local/lib/python3.10/site-packages (from hassil==0.2.3->-r requirements.txt (line 1)) (4.11.1)
Requirement already satisfied: pathspec>=0.9.0 in /home/vscode/.local/lib/python3.10/site-packages (from black==22.12.0->-r requirements.txt (line 5)) (0.10.3)
Requirement already satisfied: tomli>=1.1.0 in /home/vscode/.local/lib/python3.10/site-packages (from black==22.12.0->-r requirements.txt (line 5)) (2.0.1)
Requirement already satisfied: click>=8.0.0 in /home/vscode/.local/lib/python3.10/site-packages (from black==22.12.0->-r requirements.txt (line 5)) (8.1.3)
Requirement already satisfied: mypy-extensions>=0.4.3 in /home/vscode/.local/lib/python3.10/site-packages (from black==22.12.0->-r requirements.txt (line 5)) (0.4.3)
Requirement already satisfied: platformdirs>=2 in /home/vscode/.local/lib/python3.10/site-packages (from black==22.12.0->-r requirements.txt (line 5)) (2.6.2)
Requirement already satisfied: pyflakes<3.1.0,>=3.0.0 in /home/vscode/.local/lib/python3.10/site-packages (from flake8==6.0.0->-r requirements.txt (line 6)) (3.0.1)
Requirement already satisfied: mccabe<0.8.0,>=0.7.0 in /home/vscode/.local/lib/python3.10/site-packages (from flake8==6.0.0->-r requirements.txt (line 6)) (0.7.0)
Requirement already satisfied: pycodestyle<2.11.0,>=2.10.0 in /home/vscode/.local/lib/python3.10/site-packages (from flake8==6.0.0->-r requirements.txt (line 6)) (2.10.0)
Requirement already satisfied: typing-extensions>=3.10 in /home/vscode/.local/lib/python3.10/site-packages (from mypy==0.991->-r requirements.txt (line 8)) (4.4.0)
Requirement already satisfied: dill>=0.2 in /home/vscode/.local/lib/python3.10/site-packages (from pylint==2.15.9->-r requirements.txt (line 9)) (0.3.6)
Requirement already satisfied: astroid<=2.14.0-dev0,>=2.12.13 in /home/vscode/.local/lib/python3.10/site-packages (from pylint==2.15.9->-r requirements.txt (line 9)) (2.13.2)
Requirement already satisfied: tomlkit>=0.10.1 in /home/vscode/.local/lib/python3.10/site-packages (from pylint==2.15.9->-r requirements.txt (line 9)) (0.11.6)
Requirement already satisfied: pluggy<2.0,>=0.12 in /home/vscode/.local/lib/python3.10/site-packages (from pytest==7.2.0->-r requirements.txt (line 10)) (1.0.0)
Requirement already satisfied: exceptiongroup>=1.0.0rc8 in /home/vscode/.local/lib/python3.10/site-packages (from pytest==7.2.0->-r requirements.txt (line 10)) (1.1.0)
Requirement already satisfied: iniconfig in /home/vscode/.local/lib/python3.10/site-packages (from pytest==7.2.0->-r requirements.txt (line 10)) (2.0.0)
Requirement already satisfied: attrs>=19.2.0 in /home/vscode/.local/lib/python3.10/site-packages (from pytest==7.2.0->-r requirements.txt (line 10)) (22.2.0)
Requirement already satisfied: packaging in /home/vscode/.local/lib/python3.10/site-packages (from pytest==7.2.0->-r requirements.txt (line 10)) (23.0)
Requirement already satisfied: lazy-object-proxy>=1.4.0 in /home/vscode/.local/lib/python3.10/site-packages (from astroid<=2.14.0-dev0,>=2.12.13->pylint==2.15.9->-r requirements.txt (line 9)) (1.9.0)
Requirement already satisfied: wrapt<2,>=1.11 in /home/vscode/.local/lib/python3.10/site-packages (from astroid<=2.14.0-dev0,>=2.12.13->pylint==2.15.9->-r requirements.txt (line 9)) (1.14.1)
@poltalashka ➜ /workspaces/intents (mybn3main) $
```

**Reply by arunshekher** [reply]
@poltalashka Basically we are adopting test-driven development.

- We create syntax for intent sentences (meant to control entities) in sentences files with `<slots>` for these entities.
- We test the `intent sentences` against an expected asserted _Bengali language sentence_ in the test file of the same name (slots swapped-out with some fake entity names) as the sentence file.

Sharing all the source files I used to test:

`sentences/bn/homeassistant_HassTurnOff.yaml`
```yaml
language: bn
intents:
  HassTurnOff:
    data:
      - sentences:
          - "বন্ধ করুন <name>"
```

`tests/bn/homeassistant_HassTurnOff.yaml`
```yaml
language: bn
tests:
  - sentences:
      - "বন্ধ করুন শয়নকক্ষ বাতি"
    intent:
      name: HassTurnOff
      slots:
        name: light.bedroom_lamp
```

`sentences/bn/_common.yaml`
```yaml
language: bn
responses:
  errors:
    no_intent: “দুঃখিত আমি বুঝতে পারিনি”
    no_area: “{{ area }} এর ক্ষেত্র আগে উঠানো হয়নি ”
    no_domain: “{{ area }} এরমধ্যে  {{ domain }} নেই”
    no_device_class: “{{ area }} এরমধ্যে  {{ device_class }} নেই
    no_entity: “এই নামে কোন মেশিন বা  {{ entity }} নামে  কিছু নেই”
    handle_error: “কাজটি শেষ করার আগে অপ্রত্যাশিত একটি ভুলসংঘটিত হয়েছে”
lists: {}
expansion_rules:
  name: "{name}"
skip_words: []
````
`tests/bn/_fixtures.yaml`
```yaml
language: bn
areas:
  - name: Kitchen
    id: kitchen
  - name: Living Room
    id: living_room
  - name: Bedroom
    id: bedroom
  - name: Garage
    id: garage
entities:
  - name: "শয়নকক্ষ বাতি"
    id: light.bedroom_lamp
    area: bedroom
  - name: Kitchen Switch
    id: switch.kitchen
    area: kitchen
  - name: Ceiling Fan
    id: fan.ceiling
    area: living_room
```

Try running the test with `pytest tests --language bn -k homeassistant_HassTurnOff` command afterwords.
Hope this helps.

**Reply by arunshekher** [reply]
@poltalashka Since you are also using codespaces I don't think you need to update the requirement manually (I assumed you are working on forked and locally cloned repo). 

But your `requirement.txt ` looks definitely outdated it doesn't have `regex==2021.8.28` PyPI package. 

I think it's recommended to save your work externally and start a new `codespace` with the current state of the repo and add your work there.

This is how my `requirements.txt` looks:

```text
hassil==0.2.4
PyYAML==6.0
voluptuous==0.13.1
regex==2021.8.28
jinja2==3.1.2

black==22.12.0
flake8==6.0.0
isort==5.11.4
mypy==0.991
pylint==2.15.10
pytest==7.2.1
```
PS: I took the liberty to format the log you posted in your post above, hope you don't mind. Apologies if that wasn't necessary.

**Reply by poltalashka** [reply]
I just started codespace with new branch bn4.main, and the requirements.txt still looks obsolete, regex lib version is not specified among other things. 

![image](https://user-images.githubusercontent.com/42247336/215235455-11c125d3-b930-4b12-a1e7-5fafbde0da59.png)

**Reply by poltalashka** [reply]
@arunshekher, I left requirements.txt as-is but replicated your work in explanation in the new branch. It worked as you documented. **Thanks for that! ** I can build on top of this now. Finally some progress...

**Comment by arunshekher** [comment]
@poltalashka glad to know the progress. 

But what I meant was to discard/leave-alone the old codespace container and _**start a new codespace**_ container. 

<img width="1001" alt="Screen Shot 2023-01-28 at 2 20 37 PM" src="https://user-images.githubusercontent.com/315195/215261576-4debda4c-a5b3-4dba-b318-3358e5de509a.png">

Or [rebuild the container persisting existing data](https://docs.github.com/en/codespaces/codespaces-reference/performing-a-full-rebuild-of-a-container).

Or if you still want to use the old container do a :
```shell
git stash
git pull origin main
git stash apply
```
in the current working branch.

**Reply by arunshekher** [reply]
@poltalashka, if I can be of any other help please don't hesitate to ask. I am ready to help as much as I can.

**Reply by poltalashka** [reply]
@synesthesiam ,, PR published [https://github.com/home-assistant/intents/pull/820](url) for HA_HassTurnOn/Off-yaml.

Thank-you @arunshekher for the details.

**Reply by arunshekher** [reply]
@poltalashka saw your PR (#820). Great job! Just want to point out that the link in your post is broken.

**Comment by andreasbrett** [comment]
Not sure why @poltalashka didn't make language leader back in 2023. Wanted to ask @DrHaque if he wants to become [language leader](https://developers.home-assistant.io/docs/voice/language-leaders/). He posted 2 PRs for Bengali already and is eligible for the role of language leader.

**Comment by poltalashka** [comment]
that's fine I can work with @DrHaque <https://github.com/DrHaque>.  Count
Me in as a team member if you haven't already.
Thx

**Reply by DrHaque** [reply]
@andreasbrett @poltalashka Thank you. Sure, I can take the Bengali leader role. I will try to get more Bengali translations done in the next few weeks.

**Reply by andreasbrett** [reply]
Great, I've prepared https://github.com/OHF-Voice/intents/pull/4039 which should be reviewed soon. Thanks for your contributions and looking forward to seeing Bengali being represented more in HA Voice!
